### ISSUE TITLE
Exporter configuration issues

### ISSUE THREAD

**Issue by rickjames961** [martin-helmich/prometheus-nginxlog-exporter#34]
**Issue type**:
> Support request

**What happened**:
when i attempt to run the exporter from the command line, i appear to have some issues with the port configuration. I do not want to hard code an ip in case the ip of the server changes, but i dont see much information on configuring the exporter accordingly. i am getting the error 

```
./prometheus-nginxlog-exporter -config-file /etc/prometheus/nginx-log-exporter.hcl
loading configuration file /etc/prometheus/nginx-log-exporter.hcl
using configuration {{%!s(int=4040) 127.0.0.1} {%!s(bool=true) 127.0.0.1:8500 dc1 http  {nginx-exporter nginx-exporter [foo bar]}} [{app-1 [/jet/log/nginx/access.log] $remote_addr - $remote_user [$time_local] "$request" $status $body_bytes_sent "$http_referer" "$http_user_agent" map[environment:beta foo:bar app:application-one] [] [] []}] %!s(bool=false)}
registering service in Consul
panic: Unexpected response code: 404 
```

Here is my hcl config
```
listen {
  port = 4040
  address = "127.0.0.1"
}

consul {
  enable = true
  address = "127.0.0.1:8500"
  datacenter = "dc1"
  scheme = "http"
  token = ""
  service {
    id = "nginx-exporter"
    name = "nginx-exporter"
    tags = ["foo", "bar"]
  }
}

namespace "app-1" {
  format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\""
  source_files = [
    "/jet/log/nginx/access.log"
  ]
  labels {
    app = "application-one"
    environment = "beta"
    foo = "bar"
  }
}
```

and nginx config
```
server {
    listen       80 default_server;
    listen       [::]:80;
    listen 	 [::]:1;
    listen       4040;
    listen       8500;

    server_name  _;
    root         /jet/app/wordpress;

    index index.php index.html index.htm;
    autoindex on;

    location /status/format/json {
        stub_status on;

        access_log on;
        allow 127.0.0.1;
        allow 98.251.250.215;
        deny all;
    }
    # Load configuration files for the default server block.
    include /jet/etc/nginx/conf.d/*.inc;
 
}
```

**Comment by martin-helmich** [comment]
Why have you configured your NGINX to listen on ports 4040 and 8500? Two issues with this:

1. Both the exporter and NGINX are configured to listen on port 4040. That's bound to crash. Pick one (in the default configuration, the exporter is configured to listen on port 4040; NGINX should not touch that port).
2. The exporter expects a Consul server to be running at `127.0.0.1:8500` (because it's in your configuration file -- remove the `consul` section if you don't need Consul integration), when in reality it's just your Nginx listening at that port responding with a `404 Not Found`. Actually, you probably don't need your Nginx to listen on port 8500 at all.

Long story short: Don't have NGINX listen at ports 4040 and 8500. If you don't have a Consul server running, remove the `consul` section from the config.

**Comment by rickjames961** [comment]
Thank you, i was able to get that fixed. I am now getting the following
issue related to the namespace?

[root@ip-10-0-139-129 bin]# ./prometheus-nginxlog-exporter -config-file
/etc/prometheus/nginx-log-exporter.hcl

loading configuration file /etc/prometheus/nginx-log-exporter.hcl

using configuration {{%!s(int=4040) localhost} {%!s(bool=false)
127.0.0.1:8500 dc1 http  {nginx-exporter nginx-exporter [foo bar]}} [{app-1
[/jet/log/nginx/access.log] $remote_addr - $remote_user [$time_local]
"$request" $status $body_bytes_sent "$http_referer" "$http_user_agent"
map[environment:beta foo:bar app:application-one] [] [] []}]
%!s(bool=false)}

starting listener for namespace app-1

running HTTP server on address localhost:4040

panic: descriptor Desc{fqName: "app-1_http_response_count_total", help:
"Amount of processed HTTP requests", constLabels: {}, variableLabels: [app
environment foo method status]} is invalid:
"app-1_http_response_count_total" is not a valid metric name


goroutine 8 [running]:

github.com/martin-helmich/prometheus-nginxlog-exporter/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc420046440,
0xc420110770, 0x1, 0x1)

/home/travis/gopath/src/
github.com/martin-helmich/prometheus-nginxlog-exporter/vendor/github.com/prometheus/client_golang/prometheus/registry.go:404
+0x9e

github.com/martin-helmich/prometheus-nginxlog-exporter/vendor/github.com/prometheus/client_golang/prometheus.MustRegister(0xc420110770,
0x1, 0x1)

/home/travis/gopath/src/
github.com/martin-helmich/prometheus-nginxlog-exporter/vendor/github.com/prometheus/client_golang/prometheus/registry.go:152
+0x53

main.(*Metrics).Init(0xc420046cc0, 0xc42003bf58)

/home/travis/gopath/src/
github.com/martin-helmich/prometheus-nginxlog-exporter/main.go:103 +0x9c2

main.main.func2(0xc420014ca9, 0x5, 0xc420045f80, 0x1, 0x1, 0xc420048070,
0x70, 0xc42000ff50, 0x0, 0x0, ...)

/home/travis/gopath/src/
github.com/martin-helmich/prometheus-nginxlog-exporter/main.go:177 +0x8d

created by main.main

/home/travis/gopath/src/
github.com/martin-helmich/prometheus-nginxlog-exporter/main.go:173 +0x5db

**Comment by martin-helmich** [comment]
Namespace names are not allowed to contain slashes. See #11 for more
information.

**Comment by rickjames961** [comment]
Awseome. Ive got that figured out and exporting metrics, i just have one
more question. I am following your post here
https://www.martin-helmich.de/en/blog/monitoring-nginx.html and the metrics
you are showing at the bottom are not included with whats being exported. I
really appreciate your help, this was the best option to get performance
metrics on my nginx server. I tried installing one of the nginx modules and
broke everything. Here is an output of the metrics i am getting.

# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 9.86e-06
go_gc_duration_seconds{quantile="0.25"} 1.1551e-05
go_gc_duration_seconds{quantile="0.5"} 1.1981e-05
go_gc_duration_seconds{quantile="0.75"} 1.28e-05
go_gc_duration_seconds{quantile="1"} 3.3923e-05
go_gc_duration_seconds_sum 0.004453872
go_gc_duration_seconds_count 339
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 21
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 2.552872e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated,
even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 1.65582496e+08
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the
profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.451156e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 635320
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage
collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 405504
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and
still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 2.552872e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 2.760704e+06
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 3.203072e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 8931
# HELP go_memstats_heap_released_bytes_total Total number of heap
bytes released to OS.
# TYPE go_memstats_heap_released_bytes_total counter
go_memstats_heap_released_bytes_total 0
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 5.963776e+06
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970
of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.5203452518495688e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 1039
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 644251
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache
structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 1736
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache
structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 16384
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 26448
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan
structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 81920
# HELP go_memstats_next_gc_bytes Number of heap bytes when next
garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 4.194304e+06
# HELP go_memstats_other_sys_bytes Number of bytes used for other
system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 476260
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the
stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 327680
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from
system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 327680
# HELP go_memstats_sys_bytes Number of bytes obtained by system. Sum
of all system allocations.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 8.72268e+06
# HELP http_request_duration_microseconds The HTTP request latencies
in microseconds.
# TYPE http_request_duration_microseconds summary
http_request_duration_microseconds{handler="prometheus",quantile="0.5"} 1160.169
http_request_duration_microseconds{handler="prometheus",quantile="0.9"} 2158.168
http_request_duration_microseconds{handler="prometheus",quantile="0.99"}
2271.968
http_request_duration_microseconds_sum{handler="prometheus"} 187710.012
http_request_duration_microseconds_count{handler="prometheus"} 126
# HELP http_request_size_bytes The HTTP request sizes in bytes.
# TYPE http_request_size_bytes summary
http_request_size_bytes{handler="prometheus",quantile="0.5"} 169
http_request_size_bytes{handler="prometheus",quantile="0.9"} 169
http_request_size_bytes{handler="prometheus",quantile="0.99"} 169
http_request_size_bytes_sum{handler="prometheus"} 23270
http_request_size_bytes_count{handler="prometheus"} 126
# HELP http_requests_total Total number of HTTP requests made.
# TYPE http_requests_total counter
http_requests_total{code="200",handler="prometheus",method="get"} 126
# HELP http_response_size_bytes The HTTP response sizes in bytes.
# TYPE http_response_size_bytes summary
http_response_size_bytes{handler="prometheus",quantile="0.5"} 1707
http_response_size_bytes{handler="prometheus",quantile="0.9"} 1713
http_response_size_bytes{handler="prometheus",quantile="0.99"} 1715
http_response_size_bytes_sum{handler="prometheus"} 214029
http_response_size_bytes_count{handler="prometheus"} 126
# HELP nginx_http_response_count_total Amount of processed HTTP requests
# TYPE nginx_http_response_count_total counter
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="GET",status="200"}
238
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="GET",status="302"}
4
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="GET",status="304"}
91
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="GET",status="404"}
20
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="GET",status="499"}
1
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="HEAD",status="200"}
3023
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="HEAD",status="404"}
1
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="POST",status="200"}
1430
nginx_http_response_count_total{app="application-one",environment="beta",foo="bar",method="POST",status="499"}
1
# HELP nginx_http_response_size_bytes Total amount of transferred bytes
# TYPE nginx_http_response_size_bytes counter
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="GET",status="200"}
1.8335801e+07
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="GET",status="302"}
38
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="GET",status="304"}
0
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="GET",status="404"}
115870
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="GET",status="499"}
0
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="HEAD",status="200"}
0
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="HEAD",status="404"}
0
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="POST",status="200"}
6.24781e+06
nginx_http_response_size_bytes{app="application-one",environment="beta",foo="bar",method="POST",status="499"}
0
# HELP nginx_parse_errors_total Total number of log file lines that
could not be parsed
# TYPE nginx_parse_errors_total counter
nginx_parse_errors_total 0
# HELP process_cpu_seconds_total Total user and system CPU time spent
in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 11.77
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1024
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 9
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 1.04448e+07
# HELP process_start_time_seconds Start time of the process since unix
epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.52030765052e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.4934016e+07

**Comment by martin-helmich** [comment]
From the [FAQ](https://github.com/martin-helmich/prometheus-nginxlog-exporter#frequently-asked-questions):

>Most metrics require certain values to be present in the access log files that are not present in the NGINX default configuration. Especially, make sure that the access log contains the `$upstream_response_time`, `$request_time` and/or `$body_bytes_sent` variables. These need to be enabled in the NGINX configuration (more precisely, the `log_format` setting) and then added to the format specified for the exporter.

**Comment by rickjames961** [comment]
I saw that in the FAQ. Here are my configurations. What am i missing?
logexporter.hcl

listen {

  port = 4040

  address = "0.0.0.0"

}


consul {

  enable = false

  address = "127.0.0.1:8500"

  datacenter = "dc1"

  scheme = "http"

  token = ""

  service {

    id = "nginx-exporter"

    name = "nginx-exporter"

    tags = ["foo", "bar"]

  }

}


namespace "nginx" {

  format = "$remote_addr - $remote_user [$time_local] \"$request\" $status
$body_bytes_sent \"$http_referer\" \"$ht

tp_user_agent\" $request_time $upstream_connect_time $upstream_header_time
$upstream_response_time"

  source_files = [

    "/jet/log/nginx/access.log"

  ]

  labels {

    app = "application-one"

    environment = "beta"

    foo = "bar"

  }

}

nginx.conf


    log_format upstream_time '$remote_addr - $remote_user [$time_local] '

                             '"$request" $status $body_bytes_sent '

                             '"$http_referer" "$http_user_agent" '

                             'rt=$request_time uct="$upstream_connect_time"
uht="$upstream_header_time" urt="$upstream_response_time"';


    access_log  /jet/log/nginx/access.log  upstream_time;

**Comment by martin-helmich** [comment]
Your log formats do not match. You have to customize the exporter's config file to your specific configuration (in this case, concerning NGINX' `log_format` directive and the exporter's `format`).
